packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,118 @@ impl<'m> RunnerImpl<'m> {
+                    inner_field,
                     None,
-                )?])
-            }
-            DataType::FixedSizeList(field, multiplier) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+                )?);
+            },
+            DataType::FixedSizeList(inner_field, size) => {
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
-                    Some(data.len * multiplier as usize),
-                )?])
+                    inner_field,
+                    Some(data.len * *size as usize),
+                )?);
             }
-            DataType::Struct(fields) => {
-                let mut v = Vec::new();
-                for (i, field) in fields.iter().enumerate() {
+            DataType::Struct(inner_fields) => {
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                for (i, inner_field) in inner_fields.iter().enumerate() {
                     let child = child_data.get(i as u32)?;
-                    v.push(self.array_data_from_js(
+                    builder = builder.add_child_data(self.array_data_from_js(
                         mv8,
                         &child,
-                        field.data_type(),
+                        inner_field,
                         Some(data.len),
                     )?);
                 }
-                Ok(v)
             }
-            t => {
-                if n_children == 0 {
-                    Ok(vec![])
-                } else {
-                    Err(Error::FlushType(t))
-                }
-            }
-        }?; // TODO: More types?
-
-        // TODO: Extra copies (in `new_buffer`) of buffers here,
-        //       because JS Arrow doesn't align things properly.
-        //       (Due to which buffer capacities are currently unused.)
-
-        // This target length is used because the JS repr does not mirror
-        // buffer building as Rust Arrow and pyarrow.
-        let target_len = len.unwrap_or(data.len);
-
-        let null_bit_buffer = if data.null_bits_ptr.is_null() {
-            None // Can't match on `std::ptr::null()`, because not compile-time const.
-        } else {
-            let capacity = data.null_bits_capacity;
-            // Ceil division.
-            let n_bytes = (target_len / 8) + (if target_len % 8 == 0 { 0 } else { 1 });
-            Some(unsafe { self.new_buffer(data.null_bits_ptr, n_bytes, capacity) })
-            // Some(unsafe { Buffer::from_unowned(data.null_bits_ptr, n_bytes, capacity) })
+            DataType::FixedSizeBinary(size) => unsafe {
+                let values = slice::from_raw_parts(data.buffer_ptrs[0], data.len * *size as usize);
```

Comment:
```suggestion
                let values = slice::from_raw_parts(data.buffer_ptrs[0], *size as usize * data.len);
```
Slightly easier to parse